### PR TITLE
Test fixes

### PR DESCRIPTION
**Pull request by bsdjhb**
The goal of this series is to get riscv64 kyua tests running to completion without panicking the kernel for various combinations of purecap and hybrid kernels and worlds.  So far I'm focused on hybrid worlds with purecap kernels as that is the one that seems to fall over in our weekly tests.

### CODE REVIEW

**Comment by bsdjhb** [comment]
This allows the hybrid kyua test suite to finish on a purecap RISC-V kernel now.

```
5969/6132 passed (163 failed)
```

**Comment by bsdjhb** [comment]
Hybrid world on hybrid RISC-V kernel:

```
6001/6132 passed (131 failed)
```

**Comment by bsdjhb** [comment]
Purecap world on hybrid kernel:

```
5978/6107 passed (129 failed)
```

File: sys/cheri/cheri_cap_relocs.c [hunk 1]
```diff
@@ -60,6 +64,37 @@ init_linker_file_cap_relocs(const void *start_relocs, const void *stop_relocs,
...
-	    data_cap, can_set_code_bounds, base_addr);
+
+	/*
+	 * This cannot use cheri_init_globals_impl directly as symbols
+	 * for kernel modules in the vnet and dpcpu sets need to use
+	 * alternate base addresses and capabilities.  Instead, we
+	 * invoke a caller-supplied callback on each capability to
+	 * request the base address and source capability.
+	 */
+	for (const struct capreloc *reloc = start_relocs;
```
Comment (arichardson) [inline sys/cheri/cheri_cap_relocs.c:75]: While this is not a lot of code duplication, I'd prefer if we could make the compiler-provided header more flexible. I think adding it here is fine as long as we have an issue in llvm-project tracking this missing feature.

Comment (bsdjhb) [reply sys/cheri/cheri_cap_relocs.c:75]: I had talked about that with Jess on IRC a bit.  I think one could make the "core" function in the header take a callback and have the existing functions construct a struct with the three caps (pc, data, rodata) and use a default callback that uses that struct for existing API functions.

**Comment by bsdjhb** [comment]
Purecap world on purecap kernel:

```
5972/6107 passed (135 failed)
```

File: sys/riscv/riscv/unwind.c [hunk 2]
```diff
@@ -50,6 +50,15 @@ unwind_frame(struct thread *td, struct unwind_state *frame)
 	if (!kstack_contains(td, fp - sizeof(fp) * 2, sizeof(fp) * 2))
 		return (false);
 
+#ifdef __CHERI_PURE_CAPABILITY__
+	if ((ptraddr_t)(fp - sizeof(fp) * 2) < cheri_getbase(fp) ||
```
Comment (arichardson) [inline sys/riscv/riscv/unwind.c:54]: Does it make sense to add this check to kstack_contains or do we also use that for invalid pointers?

Comment (bsdjhb) [reply sys/riscv/riscv/unwind.c:54]: kstack_contains uses the td_kstack members for its bounds check which may or may not match the bounds of `fp` itself (which is the thing we are about to dereference).  In particular, if we have a stack with an unexpected frame (which is the case that I ran into) you might have a `fp` cap that is within the bounds of td_kstack, but can't be dereferenced (e.g. a spilled bounded stack pointer or some such).

**Pull request merged**